### DISCUSSION TITLE
Illegal Character in File Name when saving

### DISCUSSION THREAD

**Discussion by dnaldoog** [RomanKubiak/ctrlr#612] [General]
I wanted to have a situation where a panel saves a sysex dump to file but offers the patch name as the filename in the save file dialog.

Often you will see solutions like:

` fileToWrite = utils.saveFileWindow("Save file", File.getSpecialLocation(File.userDesktopDirectory), "*.syx", true)` 

But then the user must enter a filename in the prompt.

You can have a suggested file name appear in the 'save as field' with:

<pre>
local filetitle="CLS 222 &lt;Mod&gt; 2" -- a patch name on the Roland JD-990!
fileToWrite = utils.saveFileWindow("Save data dump as sysex", File(filetitle), "*.syx", true)`
</pre>

but this will not work with an [illegal](https://www.mtu.edu/umc/services/websites/writing/characters-avoid/)  name such as  __"CLS 222 &lt;Mod&gt; 2__. (_for Windows at least_)

So usually we would need a lot of fancy _lua_ parsing code to sanitise the name before writing to disk,  but there is a simpler solution.

<pre>
local filetitle="CLS 222 &lt;Mod&gt; 2" -- a patch name on the Roland JD-990!
fileToWrite = utils.saveFileWindow("Save data dump as sysex", File(File.createLegalPathName(filetitle)), "*.syx", true)`
</pre>

This will save as __CLS 222 Mod 2.syx__

Hope it helps.

----
[Source/Lua/JuceClasses/LCore.cpp](https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/JuceClasses/LCore.cpp)

**Comment by damiensellier** [comment]
That's really cool, It happened to me before and had to strip down characters with string.gsub to make it a valid filename.

I just checked on the JUCE File class reference and also discovered the function addTrailingSeparator() which can be very useful too for cross platform paths where the slash at the end for certain OS : 

https://docs.juce.com/master/classFile.html#a7093f18070f9aec99e3bfd2857761514

**Reply by dnaldoog** [reply]
Hi Damien!

Yes, I think in many panels I have written quite elaborate `gsub` code to parse user input.

With `uiLabelComponent` you can automatically filter user input, but I find it's a troublesome and ugly component to deal with. Now one can more easily use something like 'AlertWindow' to get renaming input.
